Npgsql/NpgsqlLargeObjectStream.cs

```diff
@@ -0,0 +1,306 @@
+    /// Note that the OpenRead/OpenReadWrite method as well as all operations performed on this stream must be wrapped inside a database transaction.
+    /// </summary>
+    public class NpgsqlLargeObjectStream : Stream
+    {
+        NpgsqlLargeObjectManager _manager;
+        uint _oid;
+        int _fd;
+        long _pos;
+        bool _writeable;
+        bool _disposed;
+
+        private NpgsqlLargeObjectStream() { }
+
+        internal NpgsqlLargeObjectStream(NpgsqlLargeObjectManager manager, uint oid, int fd, bool writeable)
+        {
+            _manager = manager;
+            _oid = oid;
+            _fd = fd;
+            _pos = 0;
+            _writeable = writeable;
+        }
+
+        void CheckDisposed()
+        {
+            if (_disposed)
+                throw new InvalidOperationException("Object disposed");
+        }
+
+        /// <summary>
+        /// Reads up to <i>count</i> bytes from the large object into the buffer.
+        /// A return value of 0 indicates end of file.
+        /// </summary>
+        /// <param name="buffer">The buffer where read data should be stored.</param>
+        /// <param name="offset">The offset in the buffer where the first byte should be stored.</param>
+        /// <param name="count">The maximum number of bytes that should be read.</param>
+        /// <returns>How many bytes actually read, or 0 if end of file was already reached.</returns>
+        [GenerateAsync]
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (buffer == null)
+                throw new ArgumentNullException("buffer");
+            if (offset < 0)
+                throw new ArgumentOutOfRangeException("offset");
+            if (count < 0)
+                throw new ArgumentOutOfRangeException("count");
+            if (buffer.Length - offset < count)
+                throw new ArgumentException("Invalid offset or count for this buffer");
+            Contract.EndContractBlock();
+
+            CheckDisposed();
+
+            count = Math.Min(count, _manager.MaxTransferBlockSize);
+
+            var bytesRead = _manager.ExecuteFunctionGetBytes("loread", buffer, offset, count, _fd, count);
+            _pos += bytesRead;
+            return bytesRead;
+        }
+
+        /// <summary>
+        /// Reads <i>count</i> bytes from the large object. The only case when fewer bytes are read is when end of stream is reached.
+        /// </summary>
+        /// <param name="buffer">The buffer where read data should be stored.</param>
+        /// <param name="offset">The offset in the buffer where the first byte should be stored.</param>
+        /// <param name="count">The maximum number of bytes that should be read.</param>
+        /// <returns>How many bytes actually read, or 0 if end of file was already reached.</returns>
+        [GenerateAsync]
+        public int ReadAll(byte[] buffer, int offset, int count)
```

Comment:
I would stick to implementing Stream and wouldn't include this extra function. IMHO the class NpgsqlLargeObjectStream shouldn't even be public, the user should only have a Stream in their hands...